sadman Sakib Rafi

## **Importing Libraries: NumPy and Pandas**


In [34]:
import numpy as np
import pandas as pd 

## **Reading CSV Files into DataFrame Objects**
### I'm Using Regression with a Flood Prediction Dataset here

In [35]:
DataFrame=pd.read_csv('/kaggle/input/playground-series-s4e5/train.csv')
test_df=pd.read_csv('/kaggle/input/playground-series-s4e5/test.csv')

## **Generating Descriptive Statistics for DataFrame Columns**


In [36]:
DataFrame.describe().T

count           mean            std  \
id                               1117957.0  558978.000000  322726.531784   
MonsoonIntensity                 1117957.0       4.921450       2.056387   
TopographyDrainage               1117957.0       4.926671       2.093879   
RiverManagement                  1117957.0       4.955322       2.072186   
Deforestation                    1117957.0       4.942240       2.051689   
Urbanization                     1117957.0       4.942517       2.083391   
ClimateChange                    1117957.0       4.934093       2.057742   
DamsQuality                      1117957.0       4.955878       2.083063   
Siltation                        1117957.0       4.927791       2.065992   
AgriculturalPractices            1117957.0       4.942619       2.068545   
Encroachments                    1117957.0       4.949230       2.083324   
IneffectiveDisasterPreparedness  1117957.0       4.945239       2.078141   
DrainageSystems                  1117957.0       4.946893       2.072333   
CoastalVulnerability             1117957.0       4.953999       2.088899   
Landslides                       1117957.0       4.931376       2.078287   
Watersheds                       1117957.0       4.929032       2.082395   
DeterioratingInfrastructure      1117957.0       4.925907       2.064813   
PopulationScore                  1117957.0       4.927520       2.074176   
WetlandLoss                      1117957.0       4.950859       2.068696   
InadequatePlanning               1117957.0       4.940587       2.081123   
PoliticalFactors                 1117957.0       4.939004       2.090350   
FloodProbability                 1117957.0       0.504480       0.051026   

                                   min        25%         50%        75%  \
id                               0.000  279489.00  558978.000  838467.00   
MonsoonIntensity                 0.000       3.00       5.000       6.00   
TopographyDrainage               0.000       3.00       5.000       6.00   
RiverManagement                  0.000       4.00       5.000       6.00   
Deforestation                    0.000       4.00       5.000       6.00   
Urbanization                     0.000       3.00       5.000       6.00   
ClimateChange                    0.000       3.00       5.000       6.00   
DamsQuality                      0.000       4.00       5.000       6.00   
Siltation                        0.000       3.00       5.000       6.00   
AgriculturalPractices            0.000       3.00       5.000       6.00   
Encroachments                    0.000       4.00       5.000       6.00   
IneffectiveDisasterPreparedness  0.000       3.00       5.000       6.00   
DrainageSystems                  0.000       4.00       5.000       6.00   
CoastalVulnerability             0.000       3.00       5.000       6.00   
Landslides                       0.000       3.00       5.000       6.00   
Watersheds                       0.000       3.00       5.000       6.00   
DeterioratingInfrastructure      0.000       3.00       5.000       6.00   
PopulationScore                  0.000       3.00       5.000       6.00   
WetlandLoss                      0.000       4.00       5.000       6.00   
InadequatePlanning               0.000       3.00       5.000       6.00   
PoliticalFactors                 0.000       3.00       5.000       6.00   
FloodProbability                 0.285       0.47       0.505       0.54   

                                         max  
id                               1117956.000  
MonsoonIntensity                      16.000  
TopographyDrainage                    18.000  
RiverManagement                       16.000  
Deforestation                         17.000  
Urbanization                          17.000  
ClimateChange                         17.000  
DamsQuality                           16.000  
Siltation                             16.000  
AgriculturalPractices                 16.000  
Encroachments         

In [37]:
DataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1117957 entries, 0 to 1117956
Data columns (total 22 columns):
 #   Column                           Non-Null Count    Dtype  
---  ------                           --------------    -----  
 0   id                               1117957 non-null  int64  
 1   MonsoonIntensity                 1117957 non-null  int64  
 2   TopographyDrainage               1117957 non-null  int64  
 3   RiverManagement                  1117957 non-null  int64  
 4   Deforestation                    1117957 non-null  int64  
 5   Urbanization                     1117957 non-null  int64  
 6   ClimateChange                    1117957 non-null  int64  
 7   DamsQuality                      1117957 non-null  int64  
 8   Siltation                        1117957 non-null  int64  
 9   AgriculturalPractices            1117957 non-null  int64  
 10  Encroachments                    1117957 non-null  int64  
 11  IneffectiveDisasterPreparedness  1117957 non-null 

### **Observation** 
* The dataset consists of 21 columns, each storing integer values.
* One column, "FloodProbability," contains floating-point values.
* The majority of columns have the data type int64, indicating they store whole numbers.
* The "FloodProbability" column has the data type float64, suggesting it can store decimal numbers.
* The memory usage of the dataset is approximately 187.6 MB.


## **Calculating and Displaying Memory Usage Before Optimization**


In [38]:
# Calculate the memory usage of the DataFrame in megabytes (MB) by summing up memory usage of each column and converting from bytes to MB
memory_usage=DataFrame.memory_usage().sum()/1024**2
print("Memory Usage Before Optimization :{:.2f} MB".format(memory_usage))

Memory Usage Before Optimization :187.65 MB


## **Memory Optimization Function for DataFrame**


In [39]:
def memory_optimization(data_frame):

    cols = data_frame.columns
    for col in cols:
        column_type = data_frame[col].dtype  # Getting the data type of  the column  of the DataFrame
        
        if column_type != object:
            col_min=data_frame[col].min()  # Calculate the minimum value of the column 'col' in the DataFrame
            col_max=data_frame[col].max()  # Calculate the maximum value of the column 'col' in the DataFrame
            
            if str(column_type)[:3] == 'int':
                # Check if the minimum and maximum values of the column 'col' fall within the range of the int8 data type
                if col_min > np.iinfo(np.int8).min and col_max < np.iinfo(np.int8).max:
                    # Convert the values in the column 'col' of the DataFrame to the int8 data type
                    data_frame[col] = data_frame[col].astype(np.int8)
                elif col_min > np.iinfo(np.int16).min and col_max < np.iinfo(np.int16).max:
                    data_frame[col] = data_frame[col].astype(np.int16)
                elif col_min > np.iinfo(np.int32).min and col_max < np.iinfo(np.int32).max:
                    data_frame[col] = data_frame[col].astype(np.int32)
                elif col_min > np.iinfo(np.int64).min and col_max < np.iinfo(np.int64).max:
                    data_frame[col] = data_frame[col].astype(np.int64)
            else:
                if col_min > np.finfo(np.float16).min and col_max < np.finfo(np.float16).max:
                    data_frame[col] = data_frame[col].astype(np.float16)
                elif col_min > np.finfo(np.float32).min and col_max < np.finfo(np.float32).max:
                    data_frame[col] = data_frame[col].astype(np.float32)
                else:
                    data_frame[col] = data_frame[col].astype(np.float64)
        else:
            data_frame[col] = data_frame[col].astype('object')
            
            
    optimized_memory_usage = data_frame.memory_usage().sum()/1024**2
    
    print("memory usage after optimization is: {:.2f} MB".format(optimized_memory_usage))
    
    
    return optimized_memory_usage

## **Comparing Memory Usage Before and After Optimization**


In [40]:
memory_usage_BEFORE = DataFrame.memory_usage().sum()/1024**2

memory_usage_AFTER = memory_optimization(DataFrame) #

print("Memory Usage Reduce By: {:.2f}%".format(100*(memory_usage_BEFORE-memory_usage_AFTER)/memory_usage_BEFORE))

memory usage after optimization is: 27.72 MB
Memory Usage Reduce By: 85.23%


# Memory Usage Reduction Achieved: 85.23%


In [41]:
DataFrame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1117957 entries, 0 to 1117956
Data columns (total 22 columns):
 #   Column                           Non-Null Count    Dtype  
---  ------                           --------------    -----  
 0   id                               1117957 non-null  int32  
 1   MonsoonIntensity                 1117957 non-null  int8   
 2   TopographyDrainage               1117957 non-null  int8   
 3   RiverManagement                  1117957 non-null  int8   
 4   Deforestation                    1117957 non-null  int8   
 5   Urbanization                     1117957 non-null  int8   
 6   ClimateChange                    1117957 non-null  int8   
 7   DamsQuality                      1117957 non-null  int8   
 8   Siltation                        1117957 non-null  int8   
 9   AgriculturalPractices            1117957 non-null  int8   
 10  Encroachments                    1117957 non-null  int8   
 11  IneffectiveDisasterPreparedness  1117957 non-null 

### **Observation After optimizing data** 

* The dataset now occupies approximately 27.7 MB of memory, significantly reduced from the previous 187.6 MB.
* Most of the columns have been optimized to use the int8 data type, consuming less memory.
* The "id" column has been optimized to use the int32 data type, further reducing memory usage.
* The "FloodProbability" column has been optimized to use the float16 data type, maintaining precision while minimizing memory usage.
* By optimizing data types, memory usage has been reduced while preserving data integrity and analysis capabilities.

# Key Functions Used in this Process

> DataFrame.memory_usage(index=True, deep=False)
**calculates how much memory each column uses, showing the bytes.**

> np.iinfo(np.int8) **returns information about the minimum and maximum representable values for the int8 data type**

> dataframe[columns].astype(np.int8) **converts the data type of the specified columns in the DataFrame to the int8 data type.**